### PR TITLE
feat: add `create` function

### PR DESCRIPTION
**Pull request by himself65**

### CODE REVIEW

File: backend/index.ts [hunk 1]
```diff
@@ -13,9 +14,9 @@ export interface CreateAuthenticatorOptions {
     // sign count
     getSignCount(key: CryptoKey, rpID: string, credentialID: ArrayBuffer): Promise<number>
 
-    incrementSignCount(key: CryptoKey, rpID: string, credentialID: ArrayBuffer): Promise<void>
+    incrementSignCount(key: CryptoKey, rpID: string, credentialID?: ArrayBuffer | null): Promise<void>
 
-    hasCredential(rpID: string, credentialID: ArrayBuffer): Promise<boolean>
+    hasCredential(rpID: string, credentialID?: ArrayBuffer | null): Promise<boolean>
```
Comment (Jack-Works) [inline backend/index.ts]: Why this change?

Comment (himself65) [reply backend/index.ts]: why not

**Comment by Jack-Works** [comment]
playground/index.ts:4:61 - error TS2345: Argument of type '{}' is not assignable to parameter of type 'CreateAuthenticatorOptions'.
  Type '{}' is missing the following properties from type 'CreateAuthenticatorOptions': getNormalizedCreateOptions, getSignCount, incrementSignCount, hasCredential, and 2 more.

File: backend/index.ts (hunk 1 above)
Comment (Jack-Works) [reply backend/index.ts]: Please introduce why you need this change

**Comment by himself65** [comment]
> playground/index.ts:4:61 - error TS2345: Argument of type '{}' is not assignable to parameter of type 'CreateAuthenticatorOptions'.
> Type '{}' is missing the following properties from type 'CreateAuthenticatorOptions': getNormalizedCreateOptions, getSignCount, incrementSignCount, hasCredential, and 2 more.

I wont update playground in this pr

**Comment by Jack-Works** [comment]
> > playground/index.ts:4:61 - error TS2345: Argument of type '{}' is not assignable to parameter of type 'CreateAuthenticatorOptions'.
> > Type '{}' is missing the following properties from type 'CreateAuthenticatorOptions': getNormalizedCreateOptions, getSignCount, incrementSignCount, hasCredential, and 2 more.
> 
> I wont update playground in this pr

But that results in a type error in the main branch. You can provide a throw-only implementation

File: backend/authenticator/index.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    if (!signType) {
+        throw new Error('Not Support Algorithm')
+    }
+    const signParams = getSignatureParams(signType)
+    const signTarget = concatenate(antData, clientDataJsonHash)
+    const signature = await crypto.subtle.sign(signParams, privateKey, signTarget)
+    // end sign
+
+    const attestationObject = encode({
+        fmt: 'packed',
```
Comment (Jack-Works) [inline backend/authenticator/index.ts:70]: Why not 'none'?

File: backend/authenticator/index.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        id: idBuffer.toString('base64'),
+        rawId: idBuffer.buffer,
+        response: {
+            clientDataJSON: clientDataJsonBuffer,
+            attestationObject,
+        } as AuthenticatorAttestationResponse,
+        type: 'public-key',
+        getClientExtensionResults(): AuthenticationExtensionsClientOutputs {
+            throw new Error('not supported')
+        },
```
Comment (Jack-Works) [inline backend/authenticator/index.ts]: I think we should not throw here, but return an empty result since we're not supporting extensions

File: backend/publicKey/create.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    // we dont trust these parameters from upstream
+    delete options.timeout
+    delete options.rp.id
+
+    const { rpID, ...normalizedOptions } = await createOptions.getNormalizedCreateOptions()
+    if (!(await createOptions.hasCredential(rpID))) {
+        return new Promise((resolve, reject) => {
+            reject(new DOMException('NotSupportedError'))
+        })
+    }
```
Comment (Jack-Works) [inline backend/publicKey/create.ts]: You should handle residential key and non-residential key differently

File: backend/publicKey/create.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    signal?: AbortSignal,
+): Promise<PublicKeyCredential | null> {
+    // we don't trust these parameters from upstream
+    delete options.timeout
+    delete options.rp.id
+
+    const { rpID, ...normalizedOptions } = await createOptions.getNormalizedCreateOptions()
+    if (!(await createOptions.hasCredential(rpID))) {
+        throw new DOMException('NotSupportedError')
+    }
```
Comment (Jack-Works) [inline backend/publicKey/create.ts]: This will break non-residential key.

File: backend/publicKey/create.ts (hunk 4 above)
Comment (himself65) [reply backend/publicKey/create.ts]: What's the difference?

**Review by Jack-Works — APPROVED** [review]
